src/util.rs

```diff
@@ -1,24 +1,22 @@
 use rand::distributions::Alphanumeric;
-use rand::{self, Rng};
+use rand::{Rng, SeedableRng};
+use rand::rngs::SmallRng;
 use std::ffi::{OsStr, OsString};
 use std::path::{Path, PathBuf};
-use std::{io, str};
+use std::io;
 
 use crate::error::IoResultExt;
 
 fn tmpname(prefix: &OsStr, suffix: &OsStr, rand_len: usize) -> OsString {
     let mut buf = OsString::with_capacity(prefix.len() + suffix.len() + rand_len);
     buf.push(prefix);
 
-    // Push each character in one-by-one. Unfortunately, this is the only
-    // safe(ish) simple way to do this without allocating a temporary
-    // String/Vec.
-    unsafe {
-        rand::thread_rng()
-            .sample_iter(&Alphanumeric)
-            .take(rand_len)
-            .for_each(|b| buf.push(str::from_utf8_unchecked(&[b as u8])))
-    }
+    let small_rng = SmallRng::from_entropy();
+    buf.push(small_rng
+        .sample_iter(&Alphanumeric)
+        .take(rand_len)
+        .collect::<String>()
+    );
```

Comment:
We can avoid allocation by using [char::encode_utf8](https://doc.rust-lang.org/std/primitive.char.html#method.encode_utf8). 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=421dfa2eebe631075a03b06e08c6e7cf

```suggestion
    let mut tmp = [0; 1];
    let small_rng = SmallRng::from_entropy();
    small_rng
        .sample_iter(Alphanumeric)
        .take(rand_len)
        .for_each(|c| buf.push((c as char).encode_utf8(&mut tmp)));
```

Can we probably resolve performance issues in combination with putting a smallrng in a thread-local variable?